File: Datastructures/Java/ArrayList.java
import java.util.List;

class ArrayList {

    private int[] data;
    private int capacity;
    private int size;

    // ------------ Constructors --------------- //

    public ArrayList() {
        capacity = 2;
        data = new int[capacity];
        size = 0;
    }

    public ArrayList(int[] lst) {
        capacity = 2;
        data = new int[capacity];
        size = 0;

        for (int i = 0; i < lst.length; i++) {
            append(lst[i]);
        }
    }

    // --------------- Resizing ---------------- //


    
    private void resize() {
        capacity *= 2;
        int[] tmp = new int[capacity];
        for (int i = 0; i < size; i++) {
            tmp[i] = data[i];
        }

        data = tmp;
    }

    private void shrink() {
        capacity /= 2;
        int[] tmp = new int[capacity];
        for (int i = 0; i < size; i++) {
            tmp[i] = data[i];
        }

        data = tmp;
    }

    // ---------------- Adding Element ---------------- //

    public void append(int x) {
        if (size == capacity) {
            resize();
        }

        data[size] = x;
        size++; 
    }

    /*
    @param x 
    @return 
    */
    public void insert(int x, int index) {

        if (index < 0 || index > size) {
            throw new IndexOutOfBoundsException("Index " + index + " is out of bounds!");
        }

        if (size == capacity) {
            resize();
        }

        if (index == size) {
            append(x);
            return;
        }

        int[] tmp = new int[capacity];

        if (index == 0) {

            tmp[0] = x;
            for (int i = 0; i < size; i++) {
                tmp[i+1] = data[i];
            }

        } else {

            for (int i = 0; i < index; i++) {
                tmp[i] = data[i];
            }

            tmp[index] = x;

            for (int i = index; i < size; i++) {
                tmp[i+1] = data[i];
            }
        }

        data = tmp;
        size++;
        return;

    }

    // ------------------- Removing Element ----------------------- //

    public void remove(int index) {

        if (index < 0 || index >=size) {
            throw new IndexOutOfBoundsException("Index " + index + " is out of bounds!");
        }

        int[] tmp = new int[capacity];

        if (index == size-1) {

            for (int i = 0; i < size-1; i++) {
                tmp[i] = data[i];
            }

        } else if (index == 0) {

            for (int i = 0; i < size-1; i++) {
                tmp[i] = data[i+1];
            }

        } else {

            for (int i = 0; i < index; i++) {
                tmp[i] = data[i];
            }

            for (int i = index; i < size-1; i++) {
                tmp[i] = data[i+1];
            }
        }

        data = tmp;
        size--;
        return;
    }

    public void clearList() {
        data = new int[capacity];
        size = 0;
    }

    public int pop(int index) {

        if (index < 0 || index >=size) {
            throw new IndexOutOfBoundsException("Index " + index + " is out of bounds!");
        }

        int[] tmp = new int[capacity];

        if (index == size-1) {

            for (int i = 0; i < size-1; i++) {
                tmp[i] = data[i];
            }

        } else if (index == 0) {

            for (int i = 0; i < size-1; i++) {
                tmp[i] = data[i+1];
            }

        } else {

            for (int i = 0; i < index; i++) {
                tmp[i] = data[i];
            }

            for (int i = index; i < size-1; i++) {
                tmp[i] = data[i+1];
            }
        }

        int j = data[index];
        data = tmp;
        size++;
        return j;

    }

    public int pop() {
        return pop(size-1);
    }

    // ---------------------- Output -------------------- //

    public void sort() {

        int key;
        int j;

        for (int i = 0; i < size; i++) {

            key = data[i];
            j = i - 1;

            while (j >= 0 && data[j] > key) {
                data[j+1] = data[j];
                j = j - 1;

            }

            data[j+1] = key;
        }

    }

    public void reverse() {

        int[] tmp = new int[capacity];

        int j = size-1;
        for (int i = 0; i < size; i++) {
            tmp[i] = data[j];
            j--;
        }

        data = tmp;

    }

    public int getElement(int index) {

        if (index < 0 || index >=size) {
            throw new IndexOutOfBoundsException("Index " + index + " is out of bounds!");
        }
        
        return data[index];
    }

    public int getLength() {
        return size;
    }

    public ArrayList copy() {
        ArrayList tmp = new ArrayList(data);
        
        return tmp;
    }

    public void print() {
        if (size == 0) {
            System.out.println("The list is empty!");
        }

        System.out.print("[ ");
        for (int i = 0; i < size; i++) {
            System.out.print(data[i] + " ");
        }

        System.out.println("]");
    }
}

File: Datastructures/Cpp/ArrayList.cpp
#include <iostream>
#include <string>
#include <cstdlib>
#include <vector>
#include <stdexcept>


using namespace std;

class ArrayList {

    private:

        int* data;
        int capacity;
        int size;

        void capacityIncrease() {
            if (size == capacity) {
                capacity *= 2;
                int* tmp;
                tmp = new int[capacity];
                for (int i = 0; i < size; i++) {
                    tmp[i] = data[i];
                } 

                delete[] data;
                data = tmp;
            }
        }

    public:

        ArrayList() {

            capacity = 2;
            data = new int[capacity];
            size = 0;
        }

        ~ArrayList() {

            delete[] data;

        }

        void append(int x) {
            if (size == capacity) {
                capacityIncrease();
            }

            data[size] = x;
            size++;
        }

        void insert(int x, int pos) {

            if (size == capacity) {
                capacityIncrease();
            }

            if ((pos < 0) || (pos > size)) {
                throw range_error("Index Error");
            } else if (pos == size) {
                data[size] = x;
                size++;
            } else {
                int* tmp;
                tmp = new int[size+1];
                for (int i = 0; i < pos; i++) {
                    tmp[i] = data[i];
                }

                tmp[pos] = x;

                for (int i = pos; i < size; i++) {
                    tmp[i+1] = data[i];
                }

                data = tmp;
                size++;
 
            }
        }

        void remove(int pos) {

            if ((pos < 0) || (pos > size)) {
                throw range_error("Index Error");
            } else {
                int* tmp;
                tmp = new int[size];
                for (int i = 0; i < pos; i++) {
                    tmp[i] = data[i];
                }
                for (int i = pos+1; i < size; i++) {
                    tmp[i-1] = data[i];
                }

                data = tmp;
                size--;
 
            }
        }

        int& operator[] (int pos) {
            if ((pos < 0) || (pos > size)) {
                throw range_error("Index Error");
            } else {
                return data[pos];
            }
        }

        void print() {

            cout << "[ ";
            for (int i = 0; i < size; i++) {
                cout << data[i] << " ";
            }

            cout << "]" << endl;
        }

};

int main() {

    ArrayList myList;
    myList.append(1);
    myList.append(2);
    myList.append(3);
    myList.append(4);

    myList.print();

    myList.insert(9, 2);

    myList.print();

    myList.remove(2);

    myList.print();

    cout << myList[2] << endl;

    return 0;
}